daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Upgrading.scala

```diff
@@ -0,0 +1,464 @@
+      )
+
+      _ <- tryAll(existingChoices.values, checkChoice(_))
+      _ <- checkKey(templateName, template.map(_.key))
+    } yield ()
+  }
+
+  private def checkType(typ: Upgrading[Ast.Type]): Boolean = {
+    AlphaEquiv.alphaEquiv(unifyPackageIds(typ.past), unifyPackageIds(typ.present))
+  }
+
+  // TODO: Consider whether we should strip package ids from all packages in the
+  // upgrade set, not just within the pair
+  private def unifyPackageIds(typ: Ast.Type): Ast.Type = {
+    def stripIdentifier(id: Ref.Identifier): Ref.Identifier =
+      if (id.packageId == packageId.present) {
+        Ref.Identifier(packageId.past, id.qualifiedName)
+      } else {
+        id
+      }
+
+    typ match {
+      case Ast.TNat(n) => Ast.TNat(n)
+      case Ast.TSynApp(n, args) => Ast.TSynApp(stripIdentifier(n), args.map(unifyPackageIds(_)))
+      case Ast.TVar(n) => Ast.TVar(n)
+      case Ast.TTyCon(con) => Ast.TTyCon(stripIdentifier(con))
+      case Ast.TBuiltin(bt) => Ast.TBuiltin(bt)
+      case Ast.TApp(fun, arg) => Ast.TApp(unifyPackageIds(fun), unifyPackageIds(arg))
+      case Ast.TForall(v, body) => Ast.TForall(v, unifyPackageIds(body))
+      case Ast.TStruct(fields) => Ast.TStruct(fields.mapValues(unifyPackageIds(_)))
+    }
+  }
+
+  private def checkKey(
+      templateName: Ref.DottedName,
+      key: Upgrading[Option[Ast.TemplateKey]],
+  ): Try[Unit] = {
+    key match {
+      case Upgrading(None, None) => Success(());
+      case Upgrading(Some(pastKey), Some(presentKey)) => {
+        val key = Upgrading(pastKey.typ, presentKey.typ)
+        if (!checkType(key))
+          Failure(TemplateChangedKeyType(templateName, key))
+        else
+          Success(())
+      }
+      case Upgrading(Some(pastKey @ _), None) =>
+        Failure(TemplateRemovedKey(templateName, pastKey))
+      case Upgrading(None, Some(presentKey @ _)) =>
+        Failure(TemplateAddedKey(templateName, presentKey))
+    }
+  }
+
+  private def checkChoice(choice: Upgrading[Ast.TemplateChoice]): Try[Unit] = {
+    val returnType = choice.map(_.returnType)
+    if (checkType(returnType)) {
+      Success(())
+    } else {
+      Failure(ChoiceChangedReturnType(choice.present.name, returnType))
+    }
+  }
+
+  private def dataTypeOrigin(
+      moduleWithMetadata: ModuleWithMetadata,
+      name: Ref.DottedName,
+  ): UpgradedRecordOrigin = {
+    moduleWithMetadata.module.templates.get(name) match {
+      case Some(template @ _) => TemplateBody(name)
+      case None => {
+        moduleWithMetadata.choiceNameMap.get(name) match {
+          case Some((templateName, choiceName)) =>
+            TemplateChoiceInput(templateName, choiceName)
+          case None =>
+            moduleWithMetadata.variantNameMap.get(name) match {
+              case Some((dataTypeName, variantName)) =>
+                VariantConstructor(dataTypeName, variantName);
+              case _ => TopLevel(name)
```

Comment:
Is top level here a union of records and enums? Why is this separation between Variants and TopLevel needed?

Reply:
Because a variant is a union of structs - when we are doing error reporting, we need to be able to distinguish between those structs that are the result of a genuine record declaration, and those structs which were only generated for each branch of a variant declaration.

For example:

```
data VariantTypeName = This { ... } | That { ... } | These { ... }
```

generates the following LF

```
variant @serializable VariantTypeName =
  | These Main:VariantTypeName.These
  | This Main:VariantTypeName.This
  | That Main:VariantTypeName.That
record @serializable VariantTypeName.These = {x : Int64; y : Text}
record @serializable VariantTypeName.This = {x : Int64}
record @serializable VariantTypeName.That = {y : Text}
```

When we do error reporting, we don't want to say "top level record Main:VariantTypeName:These is not a valid upgrade" - that would be confusing and wrong. We want to say that VariantTypeName is not a valid upgrade because one of its variants isn't valid.